pkg/speaker/lb_controller.go

```diff
@@ -0,0 +1,153 @@
+/*
+Copyright 2020 The Kubesphere Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package speaker
+
+import (
+	"context"
+
+	"github.com/openelb/openelb/pkg/constant"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// LBReconciler reconciles a service object
+type LBReconciler struct {
+	client.Client
+	record.EventRecorder
+
+	Handler func(*corev1.Service) error
+}
+
+func SetupLBReconciler(mgr ctrl.Manager, svchandler func(*corev1.Service) error) error {
+	lb := &LBReconciler{
+		Handler:       svchandler,
+		Client:        mgr.GetClient(),
+		EventRecorder: mgr.GetEventRecorderFor("lb"),
+	}
+
+	return lb.SetupWithManager(mgr)
+}
+
+func IsOpenELBService(obj runtime.Object) bool {
+	svc, ok := obj.(*corev1.Service)
+	if !ok {
+		return false
+	}
+
+	if svc.Annotations == nil {
+		return false
+	}
+
+	if svc.Spec.Type != corev1.ServiceTypeLoadBalancer {
+		return false
+	}
+
+	if value, ok := svc.Annotations[constant.OpenELBAnnotationKey]; ok && value == constant.OpenELBAnnotationValue {
+		return true
+	}
+
+	return false
+}
+
+func (r *LBReconciler) shouldReconcileEP(e metav1.Object) bool {
+	if e.GetAnnotations() != nil {
+		if e.GetAnnotations()["control-plane.alpha.kubernetes.io/leader"] != "" {
+			return false
+		}
+	}
+
+	svc := &corev1.Service{}
+	err := r.Get(context.TODO(), types.NamespacedName{Namespace: e.GetNamespace(), Name: e.GetName()}, svc)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return false
+		}
+		return true
+	}
+
+	return IsOpenELBService(svc)
+}
+
+func (r *LBReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	p := predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			return IsOpenELBService(e.Object)
+		},
```

Comment:
Why haven't you set the `UpdateFunc`? If we create a service with LB type but without the `lb.kubesphere.io/v1alpha1: openelb` annotation. 

Then what would be the result if we modify this service to include this `lb.kubesphere.io/v1alpha1: openelb`  annotation?

Reply:
> 

If we do not set updateEventFilter, updateEvent will not be intercepted. In this way, if someone deletes `lb.kubesphere.io/v1alpha1: openelb`, we can also cancel the previous announcement record.